analytics/upress_reporting/helpers/download_data_aggregator.py

```diff
@@ -0,0 +1,142 @@
+from logger import createLog
+from model import Edition, Item, Link
+from model.postgres.item import ITEM_LINKS
+from managers import DBManager
+
+# Regexes needed to parse S3 logs
+REQUEST_REGEX = r"REST.GET.OBJECT "
+# File ID includes the file name for the pdf object
+FILE_ID_REGEX = r"REST.GET.OBJECT (.+pdf\s)"
+TIMESTAMP_REGEX = r"\[.+\]"
+REFERRER_REGEX = r"https://drb-qa.nypl.org/"
+
+
+class DownloadDataAggregator:
+    """
+    Parses S3 download logs and generates list of DownloadEvents, each corresponding 
+    to a single download request.
+    """
+
+    def __init__(self, publisher, date_range):
+        self.publisher = publisher
+        self.date_range = date_range
+
+        self.s3_client = boto3.client("s3")
+        self.bucket_name = os.environ.get("DOWNLOAD_BUCKET", None)
+        self.log_path = os.environ.get("DOWNLOAD_LOG_PATH", None)
+
+        self._setup_db_manager()
+        self.logger = createLog("download_request_parser")
+
+    def pull_download_events(self):
+        '''
+        Returns list of DownloadEvents in a given reporting period.
+        '''
+        download_events = []
+
+        for date in self.date_range:
+            folder_name = date.strftime("%Y/%m/%d")
+            batch = self._load_batch(folder_name)
+            downloads_per_day = self._parse_logs(batch)
+            download_events.extend(downloads_per_day)
+        
+        self.db_manager.session.close()
+        return download_events
+
+    def _setup_db_manager(self):
+        self.db_manager = DBManager(
+            user=os.environ.get("POSTGRES_USER", None),
+            pswd=os.environ.get("POSTGRES_PSWD", None),
+            host=os.environ.get("POSTGRES_HOST", None),
+            port=os.environ.get("POSTGRES_PORT", None),
+            db=os.environ.get("POSTGRES_NAME", None),
+        )
+        self.db_manager.generateEngine()
+        self.db_manager.createSession()
+
+    def _load_batch(self, log_folder):
+        prefix = self.log_path + log_folder + "/"
+        paginator = self.s3_client.get_paginator("list_objects_v2")
+        page_iterator = paginator.paginate(
+            Bucket=self.bucket_name, Prefix=prefix)
+        return page_iterator
+
+    def _parse_logs(self, batch):
+        '''
+        The edition title, identifier, and timestamp are parsed out of the
+        S3 server access log files for UMP download requests.
+        '''
+        downloads_in_batch = []
+
+        for log_file in batch:
+            if "Contents" not in log_file:
+                path = self._redact_s3_path(log_file["Prefix"])
+                raise DownloadParsingError(
+                    f"Log files in path {path} do not exist.")
+            else:
+                for c in log_file["Contents"]:
+                    curr_key = str(c["Key"])
+                    # log_object is a dict type
+                    log_object = self.s3_client.get_object(
```

Comment:
I think we can rename this to log_dict if we want to be explicit and remove the comment!